### PR TITLE
Type ahead feature (trie data structure part) for TC #2

### PR DESCRIPTION
**Pull request by Fnuworsu**
# Description  
- What does this PR do?  
  `Implements a trie data structure for efficient autocomplete functionality with API integration`  
- Why is it necessary or valuable?  
  `Provides fast prefix-based search capabilities for the application`  
---

# Milestones / Related Work  
- Feature or user story this contributes to: `Base data structure for TC#2 `  

---

# Test Plan  
- How was this tested?  
  `Manual testing with console logs using sample API data`  
- Screenshots or recordings:  
 
Screenshot 2025-07-15 at 4 23 16 PM

## Edge Cases Tested  
*   `Empty search term handling`
*   `Case insensitivity (all strings converted to lowercase)`
*   `Filtering out irrelevant fields (id, type)`

---

# Additional Notes  
- Reviewer requests or feedback needed: `Feedback on the data extraction approach and trie implementation efficiency`

### CODE REVIEW

File: backend/typeAhead/trie.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+    find(word) {
+        let curr = this.root
+        const res = []
+        for (const c of word) {
+            if (!curr.children[c]) {
+                break
+            }
+            curr = curr.children[c]
+            res.push(curr.words)
```
Comment (sinap-fb) [inline backend/typeAhead/trie.js:32]: how come we push the words at each node?
If the `word` is "foo", i don't want to get all the words for the first "f" node (which may include things like "fan" or "friend")

File: backend/typeAhead/trie.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            "Content-Type": "application/json"
+        },
+        body:JSON.stringify({query: "byte"})
+    })
+
+    const data = await response.json()
+    return data
+}
+
+let data = await getData()
```
Comment (sinap-fb) [inline backend/typeAhead/trie.js:60]: might be helpful to have some quick-and-dirty error handling, just in case the local server isn't running?  (or maybe it already spits out a helpful error)

File: backend/typeAhead/trie.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+const extractAllStrings = (obj) => {
```
Comment (sinap-fb) [inline backend/typeAhead/trie.js:62]: would suggest a brief comment explaining what's going on here (though i understand this is rough WIP code!)

File: backend/routes/searchRoutes/searchRoutes.js [hunk 4]
```diff
@@ -6,9 +6,9 @@ const router = express.Router()
 const prisma = new PrismaClient()
 
 // Search across all content types
-router.get('/api/search', isAuthenticated, async (req, res) => {
+router.post('/api/search',isAuthenticated, async (req, res) => {
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply backend/routes/searchRoutes/searchRoutes.js:9]: I was having issues with req.query so I switched to req.body for testing. when testing with using req.body and the `method: "GET"`, insomnia didn't allow me testing GET request with a body, so i switched to POST for testing

File: backend/typeAhead/trie.js (hunk 2 above)
Comment (Fnuworsu) [reply backend/typeAhead/trie.js:60]: @sinap-fb gotcha! implementing that real quick

File: backend/typeAhead/trie.js (hunk 1 above)
Comment (sinap-fb) [reply backend/typeAhead/trie.js:32]: ty for the clear illustration!
